resilience4j-retry/src/main/java/io/github/resilience4j/retry/internal/RetryBudgetImpl.java

```diff
@@ -0,0 +1,255 @@
+                publishRetryEvent(() -> new RetryOnSuccessEvent(getName(), currentNumOfAttempts, throwable));
+            } else {
+                succeededWithoutRetryCounter.increment();
+            }
+        }
+
+        public void onError(Exception exception) throws Throwable {
+            if (exceptionPredicate.test(exception)) {
+                lastException.set(exception);
+                throwOrSleepAfterException();
+            } else {
+                failedWithoutRetryCounter.increment();
+                publishRetryEvent(() -> new RetryOnIgnoredErrorEvent(getName(), exception));
+                throw exception;
+            }
+        }
+
+        public void onRuntimeError(RuntimeException runtimeException) {
+            if (exceptionPredicate.test(runtimeException)) {
+                lastRuntimeException.set(runtimeException);
+                throwOrSleepAfterRuntimeException();
+            } else {
+                failedWithoutRetryCounter.increment();
+                publishRetryEvent(() -> new RetryOnIgnoredErrorEvent(getName(), runtimeException));
+                throw runtimeException;
+            }
+        }
+
+        private void throwOrSleepAfterException() throws Exception {
+            int currentNumOfAttempts = numOfAttempts.incrementAndGet();
+            if (canRetry(currentNumOfAttempts)) {
+                waitIntervalAfterFailure();
+            } else {
+                failedAfterRetryCounter.increment();
+                Exception throwable = lastException.get();
+                publishRetryEvent(() -> new RetryOnErrorEvent(getName(), currentNumOfAttempts, throwable));
+                throw throwable;
+            }
+        }
+
+        private void throwOrSleepAfterRuntimeException() {
+            int currentNumOfAttempts = numOfAttempts.incrementAndGet();
+            if (canRetry(currentNumOfAttempts)) {
+                waitIntervalAfterFailure();
+            } else {
+                failedAfterRetryCounter.increment();
+                RuntimeException throwable = lastRuntimeException.get();
+                publishRetryEvent(() -> new RetryOnErrorEvent(getName(), currentNumOfAttempts, throwable));
+                throw throwable;
+            }
+        }
+
+        private void waitIntervalAfterFailure() {
+            // wait interval until the next attempt should start
+            long interval = intervalFunction.apply(numOfAttempts.get());
+            Try.run(() -> sleepFunction.accept(interval))
+                    .getOrElseThrow(ex -> lastRuntimeException.get());
+        }
+
+        /**
+         * Resolves whether the failed operation can retry.
+         *
+         * @param currentNumOfAttempts current number of attempts by this operation
+         * @return boolean whether the operation can retry
+         */
+        private boolean canRetry(final int currentNumOfAttempts) {
+            // check max number of attempts
```

Comment:
Both me and @RobWin separately noticed possible issue with this design created on top of BitBuffer.
The issue is that after `retriesPctWithOneMore >= retryThreshold` actually happens, retry will be blocked and you will you actually need one call to succeed without retries to recover from this state.

Is it your actual design choice? 
For examle Finagle actually has additional time constraints for this behaviour. https://finagle.github.io/blog/2016/02/08/retry-budgets/

That is why I was suggesting you to look through AtomicRateLimiter token bucket implementation.
For now it seems to me that correct design should be some combination of buckets with BitBuffer.